source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -289,6 +291,36 @@ void Filter::onDestroy() {
   }
 }
 
+bool Filter::checkAndApplyHeaderMutation(
+    Filters::Common::MutationRules::CheckOperation operation, absl::string_view key,
+    absl::string_view value,
+    const std::function<void(Http::LowerCaseString, absl::string_view)> mutation_func) {
+  if (config_->validateMutations() && (!Http::HeaderUtility::headerNameIsValid(key) ||
+                                       !Http::HeaderUtility::headerValueIsValid(value))) {
+    ENVOY_STREAM_LOG(trace, "Rejecting invalid header '{}':'{}'.", *decoder_callbacks_, key, value);
+    rejectResponse();
+    return false;
+  }
+  // Check header mutation is valid according to configured decoder mutation rules.
+  Http::LowerCaseString lowercase_key(key);
+  switch (config_->checkDecoderHeaderMutation(operation, lowercase_key, value)) {
+  case CheckResult::OK:
+    ENVOY_STREAM_LOG(trace, "'{}':'{}'", *decoder_callbacks_, key, value);
+    if (mutation_func != nullptr) {
+      mutation_func(lowercase_key, value);
+    }
+    break;
+  case CheckResult::IGNORE:
+    ENVOY_STREAM_LOG(trace, "Ignoring invalid header '{}':'{}'.", *decoder_callbacks_, key, value);
+    break;
+  case CheckResult::FAIL:
+    ENVOY_STREAM_LOG(trace, "Rejected invalid header '{}':'{}'.", *decoder_callbacks_, key, value);
+    rejectResponse();
+    return false;
+  }
+  return true;
+}
```

Comment:
Rather than using a `mutation_func` here, it would be better to just let the method to take a `LowerCaseString` as key and validate it. And if the method return true, let the caller to set the key/value into the header map.

Reply:
This function is just a convenience function since the validation & checking of all header mutations is the same.

It can't take the key as LowerCaseString because Http::HeaderUtility::headerNameIsValid needs to be called before packing the key into a LowerCaseString for fuzzing purposes (i.e. to prevent the fuzzer from crashing).

Also, the only part of the switch statement that isn't exactly the same across all uses of this function is what happens on CheckResult::OK, so I thought it made sense to just pass in a function to reduce the boilerplate elsewhere.